### PR TITLE
feat: Wave 4 — Documenter v1, dependency manifest, CI hardening

### PR DESCRIPTION
**Pull request by noodlemctwoodle**
## Summary

Wave 4 delivers the **Sentinel Documenter** — a read-only collector + Markdown
renderer that produces a per-workspace audit pack from a live Sentinel
deployment — plus the surrounding dependency-manifest workflow, composite GitHub
Actions, and CI/pipeline hardening. 134 commits, themed below.

### Sentinel Documenter (`Scripts/Documenter/`)
A two-stage pipeline that splits the cheap, idempotent rendering step from the
authenticated collection step.

- **Collector** (`Export-SentinelInventory.ps1`) — read-only, ~35 captures
  covering Sentinel artefacts (rules, hunting queries, parsers, workbooks,
  watchlists, automation, playbooks + MI RBAC), the ingestion layer (DCRs,
  DCEs, diagnostic settings, dedicated clusters, Data Lake), workspace
  context (tables + plans, retention, pricing, subscription, locks, policy
  assignments) and 30-day `Usage` for cost. Falls back to direct REST for
  the documented Az-module gaps (CCF connectors, Content Hub catalogue,
  full DCR JSON, settings, Lake API on `Microsoft.SentinelPlatformServices`).
- **Renderer** (`Convert-SentinelInventoryToMarkdown.ps1`) — pure file→file
  transform, no Azure dependency, so Pester drives it end-to-end without
  auth. Produces 37 Markdown sections per workspace (`00-overview.md`
  through `99-references.md`) plus an `index.md`.
- **Gap engine** (`Private/GapChecks.ps1`) — 28 `Test-*` checks dispatched
  from `best-practices.json`. Each finding gets a SENT-NNN code,
  remediation pointer, and Learn URL; every mention of a finding ID in
  any section becomes a hotlink to the corresponding row in
  `90-gap-analysis.md`.
- **Mermaid → SVG pre-render** (`Convert-MermaidToSvg.ps1`) — runs `mmdc`
  over every fenced ```mermaid block after the renderer finishes, so
  charts render identically on ADO Repos preview, ADO code-wiki,
  GitHub, and MkDocs. The bundled ADO wiki Mermaid version lags
  `xychart-beta` / `sankey-beta`.
- **Data Lake coverage** (Section 88) — full coverage of
  `Microsoft.SentinelPlatformServices/sentinelPlatformServices`
  including architecture, tier chart, retention chart, billing meters,
  capabilities, asset data. Lake state flows into the topology
  flowchart (Section 83) and the cost-flow Sankey (Section 84).
- **Production-readiness fixes** (final commit) — watchlist enumeration
  no longer pulls per-item bodies (collector runtime drops from 10+ min
  to ~2 min on watchlist-heavy workspaces); playbook MI capture handles
  UserAssigned identity shape; cost-flow Sankey collapses the long tail
  into per-source bucket nodes and scales chart height dynamically;
  mindmap labels wrap so GB values no longer clip.

### Operations docs (`Docs/Operations/`)
- `Sentinel-Documenter.md` — operator guide (run locally, run in pipeline,
  permissions for the SP, OIDC federated-credential subject, where the
  rendered docs land, gap-engine internals).
- `Documenter-Renderer-Design.md` — maintainer-facing design spec for the
  renderer (two-stage rationale, file map, conventions).
- `Sentinel-Data-Lake-Coverage.md` — reader-facing reference for what
  Section 88 captures and how it integrates with sections 83 + 84.
- `DCR-Watchlist.md`, `Dependency-Manifest.md`, `Sentinel-Drift-Detection.md`
  refreshed alongside.

### Dependency manifest + composite actions (`Scripts/Build-DependencyManifest.ps1`, `Pipelines/`)
- `Build-DependencyManifest.ps1` generates `dependencies.json` from
  Sentinel artefact metadata so downstream deployments validate
  cross-references before applying.
- New composite GitHub Actions and refreshed ADO pipelines:
  `Sentinel-Documenter.yml`, `Sentinel-Dependency-Update.yml`,
  `Sentinel-DCR-Inventory.yml`, `Sentinel-Drift-Detect.yml`. The
  `Sentinel-PR-Validation.yml` workflow gates merges on artefact
  schema + KQL parse + dependency satisfaction.

### CI / pipelines hardening
- Drift-detect workflow + Test-SentinelRuleDrift.ps1 — fires when the
  deployed analytic rule diverges from the source manifest beyond a
  semver threshold.
- Build Service git-permissions guidance expanded after a live ADO
  permissions failure.
- GitHub Copilot reference + bidirectional cross-references added.

## Test plan

- [x] Full Pester suite (`Invoke-Pester -Path ./Tests`): 194 passing,
      0 failed (Documenter renderer end-to-end against the
      `Tests/Documenter/Fixtures/sample` fixture covers every section,
      every gap-engine SENT-NNN code, every REST-helper edge case)
- [x] Manual end-to-end against a production Sentinel workspace —
      collector + renderer + Mermaid SVG pass all complete cleanly,
      every section file produced, no clipped labels in the rendered
      cost charts, gap engine fires expected SENT-NNN codes
- [x] PII audit on the entire wave4 diff — no real GUIDs, no real
      workspace/RG/tenant names, no real emails, no secrets, no
      customer IPs. Test fixtures use placeholder GUIDs
      (`00000000-…`, `aaaaaaaa-…`, etc.); `Private/Resources/*.json`
      contains only MS-published reference data with documented
      `source:` URLs.
- [x] ADO pipeline parity — `Sentinel-Documenter.yml` runs the same
      collector + renderer + Mermaid SVG pass and publishes the
      `sentinel-docs` artefact on the build summary
- [ ] GitHub Actions equivalent — workflow is in place, needs a live
      run from this PR's branch to verify (federated-credential
      subject is `repo:noodlemctwoodle/Sentinel-As-Code:ref:refs/heads/feature/wave4`
      for the validation run)

## Reviewer focus

- `Scripts/Documenter/Private/GapChecks.ps1` — the 28 SENT-NNN checks
  drive the security-posture findings; correctness of these is the
  most user-visible part of the tool
- `Scripts/Documenter/Private/Resources/best-practices.json` — the
  rule manifest mapping check function → finding row. Adding a new
  finding only requires a row here plus a `Test-*` function in
  `GapChecks.ps1`
- `Pipelines/Sentinel-Documenter.yml` — the recommended schedule is
  weekly off-hours; the workflow opens a `docs(sentinel): snapshot
  <ws> <date>` PR on private repos and publishes an artefact-only
  build on public repos
- Permissions: documenter SP needs **Microsoft Sentinel Reader** +
  **Log Analytics Reader** on the workspace, **Reader** + **Monitoring
  Reader** at subscription scope, and **Reader** on the playbook RG.
  No write permissions anywhere — every API call is a GET.

## Breaking changes

None. The renderer accepts old `_raw/` snapshot shapes via fall-through
(no `watchlist-items/` directory, no `AllTablesByCost` key, etc.) so
existing captures continue to render.

### CODE REVIEW

File: dependencies.json [hunk 1]
```diff
@@ -1127,7 +1127,7 @@
     },
     "HuntingQueries/MicrosoftGraphActivityLogs/PowershellSecurityGroupCloningDetection.yaml": {
       "tables": [
-        "MicrosoftGraphActivityLogs"
+        "AuditLogs"
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply dependencies.json]: Push-back — this comment misreads the change.

1. **Diff direction**: the manifest edit went `AuditLogs` → `MicrosoftGraphActivityLogs`, not the reverse. The `+` lines are the new state. Both hunting queries currently query `MicrosoftGraphActivityLogs` as their active table — the manifest now correctly reflects that.

2. **Hand-edit claim**: the file was not hand-edited. The commit body explicitly records that `./Scripts/Build-DependencyManifest.ps1 -Mode Generate` was run from the repo root and the regenerated output committed. The 2-line diff is what the generator produced — exactly the stale entries the PR-validation `dependency-manifest` job was failing on. The check **passed on that commit**, which it would not have done against a hand-edit (Verify mode re-runs the generator and diffs).

3. **Hard rule 3 not violated**: the rule forbids hand-editing; running the generator and committing the result is the prescribed remediation path. The commit message states this.

No revert required. Closing this comment as not-actionable.

[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply dependencies.json]: Correction to point 1 of my previous reply: I had the diff direction backwards. The hunk you reviewed (at commit `f78dcf4`) did show `MicrosoftGraphActivityLogs` → `AuditLogs` — wave4 had branched **before** main was fixed and was carrying the stale `AuditLogs` state. You were right that the state at that SHA was a regression vs main.

That state was corrected in commit `4c1c28a` (`chore(deps): regenerate dependency manifest`) which ran `Build-DependencyManifest.ps1 -Mode Generate` and brought wave4 back in line with both the YAML source-of-truth and main:

```
$ grep -A4 PowershellMaliciousOauthAppInjectionDetection dependencies.json
"HuntingQueries/MicrosoftGraphActivityLogs/PowershellMaliciousOauthAppInjectionDetection.yaml": {
  "tables": [
    "MicrosoftGraphActivityLogs"
  ]
```

So the end state matches what you wanted; the path there was the generator, not a hand-edit (points 2 and 3 of the original reply still apply).

File: Scripts/Documenter/Private/Get-AzureRetailPrice.ps1 [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+.PARAMETER OutputRoot
+    Folder where retail-prices.json should be written. Typically the workspace _raw root.
+
+.PARAMETER ServiceNames
+    Service names to filter on. Defaults to the two Sentinel-relevant services. Override
+    if you want to extend the cost calculator (e.g. include 'Storage' for archive).
+
+.OUTPUTS
+    [PSCustomObject[]] — the union of all returned price rows.
+
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Get-AzureRetailPrice.ps1:35]: Fixed in 33e94f6. `.OUTPUTS` now documents the actual returned object (FetchedAtUtc, Region, Services, and the Prices array) rather than [PSCustomObject[]].

File: Scripts/Documenter/Private/Get-SentinelCostEstimate.ps1 [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    $monthlyTotal = ($perTable | Measure-Object -Property MonthlyCost -Sum).Sum
+    if (-not $monthlyTotal) { $monthlyTotal = 0.0 }
+
+    $perTableSorted = @($perTable | Sort-Object -Property MonthlyCost -Descending)
+    $top10 = $perTableSorted | Select-Object -First 10
+
+    # Commitment-tier what-if — only meaningful for PerGB2018 workspaces.
+    $commitmentWhatIf = @()
+    $sku = ($workspace.properties.sku.name) -as [string]
+    if ($sku -eq 'PerGB2018') {
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Get-SentinelCostEstimate.ps1]: Fixed in 33e94f6. The SKU read is now guarded, so a missing or corrupt workspace.json is treated as an unknown SKU instead of throwing under StrictMode.

File: Scripts/Documenter/REFERENCES.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
+# Sentinel Documenter — References & Conventions
+
+A durable record of every API version, module, KQL query and Microsoft Learn page the
+documenter depends on. When something in Microsoft's surface area changes, this page is
+the first thing to update; the change then ripples through `Documenter.psd1`,
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/REFERENCES.md]: Fixed in 33e94f6. Em-dashes removed from REFERENCES.md (count is now zero), per the documented convention.

File: Scripts/Documenter/Private/Get-SentinelCostEstimate.ps1 [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+        if ($billable30d -le 0 -or -not $name) { continue }
+
+        $plan = if ($planByTable.ContainsKey($name)) { $planByTable[$name] } else { 'Analytics' }
+        $unit = switch ($plan) {
+            'Basic'     { $unitPrices.BasicIngestionPerGb }
+            'Auxiliary' { $unitPrices.AuxiliaryIngestionPerGb }
+            'DataLake'  { $unitPrices.DataLakeIngestionPerGb }
+            default     { $unitPrices.AnalyticsIngestionPerGb }
+        }
+        if ($benefitSet.ContainsKey($name)) { $unit = 0.0 }
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Get-SentinelCostEstimate.ps1:205]: Leaving the zeroing in place; the suggested partial waiver would mis-model the benefit. The tables in sentinel-benefit-tables.json receive the full Microsoft Sentinel free-ingestion benefit (both Log Analytics ingestion and the Sentinel premium are waived for eligible tables, per the cited billing-reduce-costs page), not a premium-only waiver, so zeroing is the correct model. The genuine caveat is that it assumes the benefit is active: added that caveat to the estimator output in 33e94f6, and SENT-019 already flags workspaces with eligible Defender plans where the benefit is not detected.

File: Scripts/Documenter/Private/Get-SentinelGap.ps1 [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        [string]$GapChecksPath
+    )
+
+    Set-StrictMode -Version Latest
+    $ErrorActionPreference = 'Stop'
+
+    if (-not (Test-Path $RulesPath))      { throw "Rules file not found: $RulesPath" }
+    if (-not (Test-Path $GapChecksPath))  { throw "GapChecks not found: $GapChecksPath" }
+    if (-not (Test-Path $InputRoot))      { throw "Input root not found: $InputRoot" }
+
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Get-SentinelGap.ps1:58]: Fixed in 33e94f6. ResourcesRoot is now validated with Test-Path alongside RulesPath, GapChecksPath and InputRoot.

File: Scripts/Documenter/Private/Invoke-SentinelRest.ps1 [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                if ($response.PSObject.Properties.Name -contains 'nextLink') {
+                    $next = $response.nextLink
+                } elseif ($response.PSObject.Properties.Name -contains 'NextPageLink') {
+                    $next = $response.NextPageLink
+                }
+                $url = $next
+            } else {
+                $accumulator.Add($response)
+                $url = $null
+            }
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Invoke-SentinelRest.ps1]: Fixed in 33e94f6. Removed the dead NextPageLink branch and the Retail Prices claims from the help and comments. Retail Prices paginates in its own client (Get-AzureRetailPrice); Invoke-SentinelRest handles ARM value/nextLink only.

File: .github/workflows/sentinel-document.yml [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+
+env:
+  PESTER_VERSION: "5.7.1"
+  YAML_VERSION:   "0.4.12"
+
+jobs:
+  document:
+    name: document
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply .github/workflows/sentinel-document.yml:95]: Fixed in 33e94f6. The job is now gated (if: repository.private == true OR event == workflow_dispatch), so scheduled runs are skipped on a public repo (no permanent daily failure), while a manual dispatch still runs and the unconditional guard fails loud on a public repo before any collection.

File: Docs/Operations/Sentinel-Documenter.md [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+### B. Public GitHub source + private ADO mirror (this repository)
+- Source-of-truth is a public GitHub repo (community contributions,
+  issues, discussion).
+- Pipeline testing is mirrored to a private Azure DevOps project.
+- The **ADO pipeline** opens PRs in the private ADO mirror; its push
+  reaches `origin` (= ADO repo on the agent) only and never touches the
+  public GitHub copy.
+- The **GitHub Actions workflow's PR step refuses to run on the public
+  repo** — it falls back to artefact-only delivery there, or fails fast
+  if the operator explicitly ticks `open-pull-request`.
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Docs/Operations/Sentinel-Documenter.md]: Fixed in 33e94f6. The topology section no longer suggests artefact-only fallback on a public repo; it now states nothing is produced from the public copy and to run from the private ADO mirror.

File: Docs/Operations/Sentinel-Documenter.md [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+>
+> The Documenter generates a folder of detailed tenant configuration —
+> workspace IDs, table names, rule details, RBAC principals, cost figures,
+> network ACLs. This information **MUST NOT** land in a public repository.
+>
+> Both pipelines (ADO and GitHub Actions) include a privacy guard that
+> refuses to commit `SecurityDocs/` unless the host repository is private.
+> If you run the GitHub Actions workflow on a public repo with the PR step
+> enabled, the run **fails fast** with an explicit error message. The
+> ADO pipeline relies on the fact that ADO repos are private by default
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Docs/Operations/Sentinel-Documenter.md]: Fixed in 33e94f6. The privacy-guard note now states SecurityDocs/ is never collected or published on a public repo regardless of the open-pull-request toggle (scheduled skipped, manual fails fast).

File: Scripts/Documenter/Convert-MermaidToImage.ps1 [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    This is an Azure DevOps-only workaround. ADO Repos' markdown preview and
+    ADO's "publish code as wiki" render ```mermaid``` fences as plain code, not
+    diagrams. ADO Wiki proper renders some Mermaid but lags the spec and drops
+    the experimental chart types we emit (xychart-beta, sankey-beta). On top of
+    that, ADO blocks inline SVG images for security, so an SVG <img> shows as a
+    broken image.
+
+    PNG sidesteps both problems — it renders on every ADO markdown surface
+    (Repos preview, code-wiki, project wiki) — so PNG is the default output.
+
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Convert-MermaidToImage.ps1]: Fixed in d883233. Added the standard file header (path + creation date), bumped `#requires -Version 7` to 7.2 to match the documented PowerShell 7.2+ baseline, and removed the em-dashes from the comment-based help.

File: Scripts/Documenter/Private/Invoke-SentinelRest.ps1 [hunk 12]
```diff
@@ -0,0 +1,199 @@
+<#
+.SYNOPSIS
+    Paginating wrapper around Invoke-AzRestMethod for Sentinel and Azure Resource Manager
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Invoke-SentinelRest.ps1:9]: Fixed in d883233. Added the standard file header (repo-relative path + creation date).

File: Scripts/Documenter/Private/Get-SentinelGap.ps1 [hunk 13]
```diff
@@ -0,0 +1,148 @@
+<#
+.SYNOPSIS
+    Gap-analysis engine. Loads the best-practices ruleset, builds an in-memory
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Get-SentinelGap.ps1:9]: Fixed in d883233. Added the standard file header.

File: Scripts/Documenter/Private/Get-SentinelCostEstimate.ps1 [hunk 14]
```diff
@@ -0,0 +1,280 @@
+<#
+.SYNOPSIS
+    Compute an estimated monthly cost for the workspace from the captured 30-day Usage,
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Get-SentinelCostEstimate.ps1:9]: Fixed in d883233. Added the standard file header.

File: Scripts/Documenter/Private/Get-EffectiveConnectors.ps1 [hunk 15]
```diff
@@ -0,0 +1,249 @@
+<#
+.SYNOPSIS
+    Synthesise an "effective connectors" view from the captured inventory.
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Get-EffectiveConnectors.ps1:9]: Fixed in d883233. Added the standard file header.

File: Scripts/Documenter/Private/Get-AzureRetailPrice.ps1 [hunk 16]
```diff
@@ -0,0 +1,113 @@
+<#
+.SYNOPSIS
+    Anonymous client for the Azure Retail Prices API with on-disk caching.
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Scripts/Documenter/Private/Get-AzureRetailPrice.ps1:9]: Fixed in d883233. Added the standard file header.

File: .github/workflows/sentinel-document.yml [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+          body: |
+            # Sentinel Documenter Snapshot — `${{ vars.SENTINEL_WORKSPACE }}`
+
+            > ⚠️ **Review only — do not merge.** Merging would commit tenant
+            > configuration to `main` permanently. Treat this PR as a viewer:
+            > read the diff, then close (or just leave it open as a rolling
+            > snapshot — it refreshes automatically).
+
+            **Workflow run:** [${{ github.run_id }}](${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }})
+            **Artefact (zip):** [sentinel-docs-${{ vars.SENTINEL_WORKSPACE }}-${{ github.run_id }}](${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}#artifacts) — full report including raw JSON.
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply .github/workflows/sentinel-document.yml]: Fixed in d883233. Removed the warning emoji and the em-dashes from the generated PR-body and commit-message strings; the do-not-merge emphasis is kept as bold text.

File: Pipelines/Sentinel-Documenter.yml [hunk 18]
```diff
@@ -0,0 +1,426 @@
...
+              {
+                echo "# Sentinel Documenter Snapshot — $(sentinelWorkspaceName)"
+                echo ""
+                echo "**Generated:** $(date -u +%Y-%m-%dT%H:%MZ)"
+                BUILD_URL="$(System.CollectionUri)$(System.TeamProject)/_build/results?buildId=$(Build.BuildId)"
+                echo "**Pipeline run:** [$(Build.BuildNumber)](${BUILD_URL})"
+                echo "**Artefact (zip):** [sentinel-docs](${BUILD_URL}&view=artifacts&type=publishedArtifacts)"
+                echo "**Full overview:** see [\`$OVERVIEW_PATH\`]($OVERVIEW_PATH) in this branch — the **Files Changed** tab on the PR shows every section diff."
+                echo ""
+                echo "> ⚠️ Review only. Do not merge — merging would commit tenant configuration to \`$(targetBranch)\`."
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Pipelines/Sentinel-Documenter.yml]: Fixed in d883233. Removed the emoji and em-dashes from the ADO PR-body banner and commit message; emphasis kept as bold text.

File: Docs/Operations/Sentinel-Data-Lake-Coverage.md [hunk 19]
```diff
@@ -0,0 +1,303 @@
+# Sentinel Data Lake — Documenter coverage
+
+What the Sentinel Documenter captures and renders for Microsoft Sentinel Data
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Docs/Operations/Sentinel-Data-Lake-Coverage.md]: Fixed in d883233. Em-dashes removed from Sentinel-Data-Lake-Coverage.md (count now zero).

File: Docs/Operations/Documenter-Renderer-Design.md [hunk 20]
```diff
@@ -0,0 +1,397 @@
+# Sentinel Documenter — renderer design spec
+
+> Maintainer-facing reference for [`Scripts/Documenter/Convert-SentinelInventoryToMarkdown.ps1`](../../Scripts/Documenter/Convert-SentinelInventoryToMarkdown.ps1).
+> The user-facing operating guide is the sibling [`Sentinel-Documenter.md`](Sentinel-Documenter.md).
+
+This doc captures the design of the renderer — what each chart is driven
+by, where every helper lives, the Mermaid-safety conventions that govern
+chart emission, and how to extend any of it without regressing.
```
[the remark this answers is not in the document]
Comment (noodlemctwoodle) [reply Docs/Operations/Documenter-Renderer-Design.md]: Fixed in d883233. Em-dashes removed from Documenter-Renderer-Design.md (count now zero).
